### PR TITLE
[FLINK-2656] Fix behavior of FlinkKafkaConsumer for out of range offsets

### PR DESCRIPTION
**Pull request by rmetzger**
The FlinkKafkaConsumer is failing with an OutOfRangeException. 
Flink users are expecting Flink's Kafka Source to behave similar to Kafka's own high level kafka consumer. The Kafka consumer has a configuration parameter for controlling the behavior in case of invalid or unknown offsets (where to start).

In this pull request, I've changed the behavior of the FlinkKafkaConsumer to now throw the OutOfRangeException. I'll set the offset according to the configuration parameter.

I've also added a test case to verify the fix.

### CODE REVIEW

File: flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/LegacyFetcher.java [hunk 1]
```diff
@@ -356,29 +357,24 @@ public void run() {
 				// make sure that all partitions have some offsets to start with
 				// those partitions that do not have an offset from a checkpoint need to get
 				// their start offset from ZooKeeper
-				
-				List<FetchPartition> partitionsToGetOffsetsFor = new ArrayList<>();
+				{
```
Comment (hsaputra) [inline flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/LegacyFetcher.java]: Just curious what is the extra "{" for?

**Comment by hsaputra** [comment]
Hi @rmetzger, could you tell bit summary what the fix should do in the description of this PR? Thanks.

File: flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/LegacyFetcher.java (hunk 1 above)
Comment (rmetzger) [inline flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/LegacyFetcher.java]: I'm opening a new scope for the `partitionsToGetOffsetsFor` variable.
I'm using another list with the same name later in the code. WIth a new scope for the operation here, I can use the same variable name later again.

**Comment by rmetzger** [comment]
Sorry Henry. I've updated the description.

**Comment by StephanEwen** [comment]
Good and critical fix. Should go to
  - `master`
  - `release-0.10.0-milestone1`
  - `0.9.2`

**Comment by fhueske** [comment]
LGTM

**Comment by rmetzger** [comment]
Thank you for the review. I'll merge it

**Pull request closed without merging**